<launch>
    <!-- Define parameters -->
    <!-- The system is designed based on <<Collective fission behavior in swarming systems with density-based interaction>> 
         The parameters is defined as follows. 
         Generally speaking, increasing the kernel function radius h, the strength of density force k_den,
         and the reference density restDensity, or decreasing the strength of repulsion force k_rep, the 
         swarming system is more likely to maintain a single cluster, and the critical group size becomes larger.-->

    <arg name="particleCount" value="16"/>       <!-- 粒子的数量，如果是10的倍数的平方，则以正方形分布 -->
    <arg name="particleInterval" value="0.5"/>  <!-- 粒子之间的初始间距 -->
    <arg name="particleVisScale" value="0.5"/>  <!-- 粒子可视化尺寸 -->
    <arg name="updateInterval" value="0.02"/>   <!-- 时间步长 -->
    <arg name="h" value="10.0"/>                <!-- kernel function radius -->
    <arg name="mass" value="1.0"/>              <!-- default MASS -->
    <arg name="threshold_dist" value="0.03"/>   <!-- 粒子与虚拟粒子的碰撞检测距离 -->
    <arg name="k_den" value="100.0"/>           <!-- k_den -->
    <arg name="k_rep" value="100.0"/>           <!-- k_rep -->
    <arg name="k_fri" value="100.0"/>           <!-- k_fri -->
    <arg name="v_max" value="1.0"/>             <!-- v_max -->
    <arg name="a_max" value="1.0"/>             <!-- a_max -->
    <arg name="restDensity" value="1000000.0"/> <!-- \rho0 -->
    <arg name="g" value="-9.8"/>                <!-- 重力,no use -->
    <arg name="use_pctrl" value="true"/>        <!-- 使用位置\速度\加速度控制，默认使用位置控制 -->

    <!-- Run the sph_particles node -->
    <node name="sph_zhang" pkg="water_swarm" type="sph_zhang" output="screen">
        <param name="sph/mass"              value="$(arg mass)"/> 
        <param name="sph/restDensity"       value="$(arg restDensity)"/>
        <param name="sph/h"                 value="$(arg h)"/> 
        <param name="sph/g"                 value="$(arg g)"/>
        <param name="sph/use_pctrl"         value="$(arg use_pctrl)"/>
        <param name="sph/particleCount"     value="$(arg particleCount)"/> 
        <param name="sph/particleInterval"  value="$(arg particleInterval)"/>
        <param name="sph/updateInterval"    value="$(arg updateInterval)"/>
        <param name="sph/particleVisScale"  value="$(arg particleVisScale)"/>
        <param name="sph/threshold_dist"    value="$(arg threshold_dist)"/>
        <param name="sph/k_den"             value="$(arg k_den)"/> 
        <param name="sph/k_rep"             value="$(arg k_rep)"/> 
        <param name="sph/k_fri"             value="$(arg k_fri)"/> 
        <param name="sph/v_max"             value="$(arg v_max)"/>
        <param name="sph/a_max"             value="$(arg a_max)"/>
    </node>

    <!-- Obs map-->
    <node pkg ="dynamic_map_generator" name ="map_pub" type ="map_pub" output = "screen" args="$(find dynamic_map_generator)/resource/tmp.pcd"/>

    <!-- RViz-->
    <node name="rviz" pkg="rviz" type="rviz" required="true" args="-d $(find so3_quadrotor_simulator)/config/rviz.rviz" />
    
</launch>